File: src/main/java/pl/training/workshop/m4/s06_inlinemethod/start/TicketPricing.java
package pl.training.workshop.m4.s06_inlinemethod.start;

import java.math.BigDecimal;
import java.time.LocalTime;

import pl.training.workshop.m4.s06_inlinemethod.Ticket;

/**
 * Start: cennik kasy. Dwa pośredniki bez znaczenia (base, addFee) i jeden hak
 * nadpisywany w OnlineTicketPricing (bookingFee). Wszystkie trzy "wyglądają" na trywialne.
 */
public class TicketPricing {
    private static final BigDecimal MORNING_REDUCTION = new BigDecimal("5.00");

    public BigDecimal total(Ticket ticket) {
        return addFee(price(ticket));
    }

    public BigDecimal price(Ticket ticket) {
        BigDecimal price = base(ticket);
        return ticket.start().isBefore(LocalTime.NOON) ? price.subtract(MORNING_REDUCTION) : price;
    }

    /** Kasa nie pobiera opłaty rezerwacyjnej. Nadpisywane w OnlineTicketPricing. */
    protected BigDecimal bookingFee() {
        return new BigDecimal("0.00");
    }

    private BigDecimal base(Ticket ticket) {
        return basePrice(ticket.format());
    }

    private BigDecimal addFee(BigDecimal price) {
        return price.add(bookingFee());
    }

    private static BigDecimal basePrice(int format) {
        return switch (format) {
            case 3 -> new BigDecimal("40.00");
            case 2 -> new BigDecimal("32.00");
            default -> new BigDecimal("25.00");
        };
    }
}

File: csharp/src/Training.Workshop/M4/S06InlineMethod/Start/TicketPricing.cs
namespace Training.Workshop.M4.S06InlineMethod.Start;

/// <summary>
/// Start: cennik kasy. Dwa pośredniki bez znaczenia (Base, AddFee) i jeden hak
/// nadpisywany w OnlineTicketPricing (BookingFee). Wszystkie trzy "wyglądają" na trywialne.
/// </summary>
public class TicketPricing
{
    private const decimal MorningReduction = 5.00m;

    public decimal Total(Ticket ticket)
    {
        return AddFee(Price(ticket));
    }

    public decimal Price(Ticket ticket)
    {
        decimal price = Base(ticket);
        return ticket.Start < new TimeOnly(12, 0) ? price - MorningReduction : price;
    }

    /// <summary>Kasa nie pobiera opłaty rezerwacyjnej. Nadpisywane w OnlineTicketPricing.</summary>
    protected virtual decimal BookingFee()
    {
        return 0.00m;
    }

    private decimal Base(Ticket ticket)
    {
        return BasePrice(ticket.Format);
    }

    private decimal AddFee(decimal price)
    {
        return price + BookingFee();
    }

    private static decimal BasePrice(int format)
    {
        return format switch
        {
            3 => 40.00m,
            2 => 32.00m,
            _ => 25.00m,
        };
    }
}
